File: homing-color-studio/src/main/resources/homing/js/hue/captains/singapura/js/homing/color/studio/ColorNavSecretaryModule.js
// =============================================================================
// ColorNavSecretaryModule — the Secretary for the Color Studio's colornav Party.
//
// A pure (state, envelope) -> Step redirect (Diligent Secretaries doctrine).
// "Just redirects": an incoming NodeSelected (from the ColorTreeWidget) is
// rebroadcast to every member as NavigateTo; the ColorListWidget reacts by
// listing the selected group's colours. No DOM, no console, no captures.
// =============================================================================

var ColorNavSecretary = {

    initial: {
        lastSelected:  null,
        recentUnknown: []
    },

    behavior: function (state, envelope) {
        var msg = envelope.message;

        switch (msg.kind) {

            case "NodeSelected": {
                return {
                    newState: {
                        lastSelected:  msg.node,
                        recentUnknown: state.recentUnknown
                    },
                    actions: [{
                        kind:    "BroadcastToMembers",
                        message: { kind: "NavigateTo", node: msg.node }
                    }]
                };
            }

            default: {
                var recent = state.recentUnknown.concat([{
                    kind: msg.kind,
                    from: envelope.from
                }]).slice(-10);
                return {
                    newState: {
                        lastSelected:  state.lastSelected,
                        recentUnknown: recent
                    },
                    actions: []
                };
            }
        }
    }
};

File: homing-color-studio/src/main/java/hue/captains/singapura/js/homing/color/studio/ColorNavSecretaryModule.java
package hue.captains.singapura.js.homing.color.studio;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The Secretary for the Color Studio's {@code colornav} Party — the colour
 * navigation bus. A pure {@code (state, envelope) -> Step} redirect: an incoming
 * {@code NodeSelected} (from the tree) is rebroadcast to every member as
 * {@code NavigateTo}. The {@code ColorListWidget} reacts by listing the selected
 * group's colours. Names no consumer — the tree never knows who reacts.
 */
public record ColorNavSecretaryModule() implements DomModule<ColorNavSecretaryModule> {

    /** The single export — a JS object with {@code initial} and {@code behavior}. */
    public record ColorNavSecretary() implements Exportable._Constant<ColorNavSecretaryModule> {}

    public static final ColorNavSecretaryModule INSTANCE = new ColorNavSecretaryModule();

    @Override
    public ImportsFor<ColorNavSecretaryModule> imports() {
        return ImportsFor.<ColorNavSecretaryModule>builder().build();
    }

    @Override
    public ExportsOf<ColorNavSecretaryModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new ColorNavSecretary()));
    }
}
